src/lib/consensus/proof_of_stake.ml

```diff
@@ -1530,6 +1540,268 @@ module Snark_transition = Coda_base.Snark_transition.Make (struct
+  in
+  (* is the consensus state from the genesis epoch? *)
+  let from_genesis_epoch =
+    Length.equal consensus_state.epoch_length Length.zero
+  in
+  if in_same_epoch || from_genesis_epoch then
+    Ok consensus_state.last_epoch_data
+  else if in_next_epoch then Ok consensus_state.curr_epoch_data
+  else Error ()
+
+let epoch_snapshot_name = function
+  | `Genesis -> "genesis"
+  | `Curr -> "curr"
+  | `Last -> "last"
+
+(* Select the correct epoch snapshot to use from local state for an epoch.
+ * The rule for selecting the correct epoch snapshot is predicated off of
+ * whether or not the first transition in the epoch in question has been
+ * finalized yet, as the local state epoch snapshot pointers are not
+ * updated until the consensus state reaches the root of the transition frontier.
+ * This function does not guarantee that the selected epoch snapshot is valid
+ * (i.e. it does not check that the epoch snapshot's ledger hash is the same
+ * as the ledger hash specified by the epoch data).
+ *)
+let select_epoch_snapshot ~(consensus_state : Consensus_state.Value.t)
+    ~local_state ~epoch ~epoch_data =
+  let open Consensus_state in
+  let open Local_state in
+  let open Epoch_data in
+  let open Epoch_ledger in
+  (* is the snapshot we need the genesis snapshot? *)
+  let is_genesis_snapshot =
+    Coda_base.Frozen_ledger_hash.equal epoch_data.ledger.hash
+      genesis_ledger_hash
+  in
+  (* are we in the next epoch after the consensus state? *)
+  let in_next_epoch =
+    Epoch.equal epoch (Epoch.succ consensus_state.curr_epoch)
+  in
+  (* has the first transition in the epoch reached finalization? *)
+  let epoch_is_finalized =
+    consensus_state.curr_epoch_data.length > Length.of_int Constants.k
+  in
+  if is_genesis_snapshot then
+    (`Genesis, Some local_state.genesis_epoch_snapshot)
+  else if in_next_epoch || not epoch_is_finalized then
+    (`Curr, local_state.curr_epoch_snapshot)
+  else (`Last, local_state.last_epoch_snapshot)
+
+type local_state_sync =
+  { snapshot_id: Local_state.snapshot_identifier
+  ; expected_root: Coda_base.Frozen_ledger_hash.t }
+[@@deriving to_yojson]
+
+let required_local_state_sync ~(consensus_state : Consensus_state.Value.t)
+    ~local_state =
+  let open Coda_base in
+  let open Consensus_state in
+  let epoch = consensus_state.curr_epoch in
+  let epoch_data =
+    (* This should not fail since we are getting epoch data for the
+     * same epoch that the consensus state is in. *)
+    select_epoch_data ~consensus_state ~epoch
+    |> Result.map_error
+         ~f:
+           (Fn.const
+              "unexpected failure while selecting epoch data from consensus \
+               state")
+    |> Result.ok_or_failwith
+  in
+  let source, _snapshot =
+    select_epoch_snapshot ~consensus_state ~local_state ~epoch ~epoch_data
+  in
+  if source = `Genesis then None
+    (* We never need to do work to have the genesis snapshot*)
```

Comment:
I think this is incorrect in a subtle way: if the `epoch_data.ledger.hash` ever becomes the `genesis_ledger_hash` after genesis, there will be bugs because we won't consider the states of the local snapshots at all. `select_epoch_snapshot` should probably only return `\`Genesis` when `epoch_length=1` and `curr_epoch_data.length <= k`.  However this sort of "ledger hash cycle" (g -> h1 -> g) can only happen if no block has a coinbase; two ledgers with different total currencies can never have the same hash.